native/src/main/java/io/ballerina/lib/data/xmldata/utils/DataUtils.java

```diff
@@ -472,52 +472,67 @@ private static BMap<BString, Object> addFields(BMap<BString, Object> input, Type
-                        key = qName.getPrefix().isBlank() ? localPart : qName.getPrefix() + ":" + localPart;
-                    }
-
-                    BMap<BString, Object>  annotationRecord = ValueCreator.createMapValue(Constants.JSON_MAP_TYPE);
-                    if (!doesNamespaceDefinedInField) {
-                        BMap<BString, Object> subRecordAnnotations = ((RecordType) referredType).getAnnotations();
-                        key = getElementName(subRecordAnnotations, key);
-                        processSubRecordAnnotation(subRecordAnnotations, annotationRecord);
-                    }
-
-                    BMap<BString, Object> subRecordValue = addFields(((BMap<BString, Object>) value), referredType);
-                    addNamespaceToSubRecord(key, namespaceAnnotRecord, subRecordValue);
-                    if (annotationRecord.size() > 0) {
-                        subRecordValue.put(annotationRecord.getKeys()[0],
-                                annotationRecord.get(annotationRecord.getKeys()[0]));
-                    }
-                    recordValue.put(StringUtils.fromString(key), subRecordValue);
-                } else if (fieldType.getTag() == TypeTags.ARRAY_TAG) {
-                    processArray(fieldType, annotations, recordValue, entry);
-                } else {
-                    addPrimitiveValue(addFieldNamespaceAnnotation(key, key, annotations, recordValue),
-                            annotations, recordValue, value);
-                }
+                processRecordField(fields.get(key).getFieldType(), annotations, recordValue, entry, key, value);
             } else {
                 recordValue.put(StringUtils.fromString(key), value);
             }
         }
         return recordValue;
     }
 
+    private static void processRecordField(Type fieldType, BMap<BString, Object> annotations,
+                                           BMap<BString, Object> recordValue, Map.Entry<BString, Object> entry,
+                                           String key, Object value) {
+        fieldType = getTypeFromUnionType(fieldType, value);
+        switch (fieldType.getTag()) {
+            case TypeTags.RECORD_TYPE_TAG -> processRecord(key, annotations, recordValue, value,
+                    (RecordType) fieldType);
+            case TypeTags.ARRAY_TAG -> processArray(fieldType, annotations, recordValue, entry);
+            case TypeTags.TYPE_REFERENCED_TYPE_TAG -> {
+                Type referredType = TypeUtils.getReferredType(fieldType);
+                if (referredType.getTag() != TypeTags.RECORD_TYPE_TAG) {
+                    processRecordField(referredType, annotations, recordValue, entry, key, value);
+                    return;
+                }
+                processTypeReferenceType(fieldType, annotations, recordValue, key, value);
+            }
+            default -> addPrimitiveValue(addFieldNamespaceAnnotation(key, key, annotations, recordValue),
```

Comment:
is always its primitive, when it comes to here?

Reply:
The name has to be change. Actually what happens in `getModifiedRecord` is we extract the namespace information and attribute detail to the field and construct a json from input value. Hence we have to handle record types, reference types, intersection and union types to capture the information stored in meta data (annotations).